workspace-update-stale: add a config knob to run automatically

torquestomp opened the pull request:
<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the commits rather than adding commits on top. Use force-push when
pushing the updated commits (`jj git push` does that automatically when you
rewrite commits). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

torquestomp commented:
Fixes issue #3820

PhilipMetzger commented:
In my opinion the PR contains three unrelated topics in the stack. As the first commit is a docs cleanup (1), the second two are moving the recovering commit logic into lib (2) (which makes it usable for the Google server) and then the actually advertised PR (3) which depends on (2). I think at least the first commit should be split out, since it has no relation to rest, the other two would be nice to have but can stay together. 

And all commits are missing a motivation, as it definitely is not obvious what they want to achieve. Quoted from contributing.md:
> We are not particularly strict about the style, but please do explain the reason for the change unless it's obvious.

torquestomp commented:
> In my opinion the PR contains three unrelated topics in the stack. As the first commit is a docs cleanup (1), the second two are moving the recovering commit logic into lib (2) (which makes it usable for the Google server) and then the actually advertised PR (3) which depends on (2). I think at least the first commit should be split out, since it has no relation to rest, the other two would be nice to have but can stay together.

The only difference between (2) and (3) is one moves stuff to lib, the other to cli_util, I don't think they're very different.  I split (1) to its own PR but it's still here too, I'm not sure how to set up dependent PRs with jj.

> 
> And all commits are missing a motivation, as it definitely is not obvious what they want to achieve. Quoted from contributing.md:
> 
> > We are not particularly strict about the style, but please do explain the reason for the change unless it's obvious.

Added some more text to the commit descriptions.

yuja commented on lib/src/workspace.rs:
```diff
@@ -461,11 +464,64 @@ pub struct LockedWorkspace<'a> {
     }
 
+    // Recover this workspace to its last known checkout.
+    pub fn create_and_check_out_recovery_commit(
```
It's odd that `LockedWorkspace` has a method that does very specific thing. Maybe this can be moved to cli_util?

torquestomp replied:
I'd prefer to keep this in lib if possible for use by extensions.  How about moving it to working_copy.rs as a free function?

yuja replied:
A free function sounds good to me, but can you add some commit description about why it should be in the lib? Ah, "the extensions" means extension commands/UI that uses jj-lib?

Also, iirc, we generally avoid adding commit description (that might have to be translated) to lib, so `"RECOVERY COMMIT FROM ...` might have to be a function parameter.

torquestomp replied:
Done, yes, and done.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -429,12 +429,13 @@ impl CommandHelper {
-                ))
+                let repo = workspace_command.repo().clone();
+                let commit = repo.store().get_commit(wc_commit_id)?;
+                Ok(StaleWorkingCopy::Snapshotted((repo, commit)))
```
Perhaps, this function can return `WorkspaceCommandHelper` instead of `(repo, wc_commit)` pair?

For reference, `for_stale_working_copy()` was a `CommandHelper` method before. 4d0f1b974600fd9c1425b68b5400d51e94515939

yuja commented on cli/src/cli_util.rs:
```diff
@@ -924,7 +932,35 @@ impl WorkspaceCommandHelper {
+                            self.write_stale_commit_stats(ui, &new_commit, stats)?;
+
+                            self.user_repo = ReadonlyUserRepo::new(repo);
+                        }
```
I think this can be simplified if auto-update-stale is handled by `CommandHelper::workspace_helper()`. Something like:
```rust
let maybe_stale = workspace_command.maybe_snapshot(ui)?;
let Some(stale_err) = maybe_stale else { return Ok(workspace_command); };
if !auto_update_stale {
   return Err(stale_err.into());
}
Self::load_stale_working_copy_and_recover_from_it(..) // this function will reload workspace and repo anyway
```

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1604,7 +1653,11 @@ impl WorkspaceCommandHelper {
+    fn snapshot_working_copy(
+        &mut self,
+        ui: &Ui,
+        update_stale: bool,
```
Instead of pass `update_stale` in to this function, `SnapshotWorkingCopyResult` can be an error type to be converted to `CommandError` by caller?

yuja commented on CHANGELOG.md:
```diff
@@ -111,6 +111,9 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+* Added the config setting `snapshot.auto-update-stale` for automatically
+  running `jj workspace update-stale` when applicable.
+
```
nit: move changelog entry to the unreleased section.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1535,6 +1535,22 @@ to the current parents may contain changes from multiple commits.
         self.commit_summary_template().format(commit, formatter)
     }
 
+    pub fn write_stale_commit_stats(
```
nit: perhaps, we can extract `print_update_stats()` or something from `update_working_copy()`. It's not important that this function doesn't print the wc parents.

(can be addressed later)

torquestomp replied:
Left a TODO for now.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -413,6 +418,40 @@ impl CommandHelper {
             })
     }
 
+    pub fn load_stale_working_copy_commit(
```
I don't think we need to rewrite `for_stale_working_copy()` not to return `WorkspaceCommandHelper` in this patch. If it returned `WorkspaceCommandHelper`, the caller can just reuse it in the next patch, I think.

If `Recovered`, the `workspace_command` is instantiated for the current op, so it can be reused. If `Snapshotted`, the caller will create new `workspace_command` anyway, so it can be done in this function.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -372,7 +372,49 @@ impl CommandHelper {
+                        )?;
+                        writeln!(
+                            ui.status(),
+                            "Automatically updated to fresh commit {}",
```
nit: maybe warning is better here? I think this is an usual event.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -855,6 +896,29 @@ pub struct WorkspaceCommandHelper {
+    }
+}
+
+impl<E> From<E> for SnapshotWorkingCopyError
+where
+    E: Into<CommandError>,
+{
+    fn from(value: E) -> Self {
+        Self::Command(value.into())
+    }
```
nit: I wouldn't add automatic conversion because the caller should be careful to not map stale wc error to `Command(_)`.

yuja replied:
Ok, I don't think it's terrible to insert `.map_err(to_other_err)?`, but I agree it would look a bit unusual. We can extract `StaleWorkingCopyError` that isn't convertible to `CommandError`, but I don't have a strong opinion about that. We can refactor it later.

(No action requested)

yuja replied:
I like it, thanks.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -372,7 +372,49 @@ impl CommandHelper {
         let mut workspace_command = self.workspace_helper_no_snapshot(ui)?;
-        workspace_command.maybe_snapshot(ui)?;
+
+        let workspace_command = match workspace_command.maybe_snapshot_impl(ui) {
```
Just an idea, but maybe we can move almost all codes from `cmd_workspace_update_stale()`?

```rust
fn cmd_workspace_update_stale() {
    let (workspace_command, was_stale) = command.recover(ui)?;
    if !was_stale {
        writeln!(ui.status(), "Nothing to do...")?;
    }
}

impl WorkspaceCommand {
    fn workspace_helper() {
        if auto_update_stale {
            let (..) = self.recover(ui)?;
            if was_stale {
               writeln!(ui.warning_default(), "Automatically recovered ...")?;
            }
        } else {
            ..
        }
    }
```

torquestomp replied:
After the rest of the refactoring it's fairly straightforward to consolidate this; done.  Did it in an extra commit at the end of the PR

torquestomp commented on lib/src/working_copy.rs:
```diff
@@ -369,6 +372,55 @@ impl WorkingCopyFreshness {
+    Reset(#[from] ResetError),
+    /// Checkout attempted to modify the root commit.
+    #[error("Cannot rewrite the root commit")]
+    RewriteRootCommit,
```
Done. I thought it was weird for a struct to hold another struct which holds nothing but it doesn't really hurt anything.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -367,7 +373,40 @@ impl CommandHelper {
+                    StaleWorkingCopyStatus::RecoveredSnapshot(id) => {
+                        writeln!(
+                            ui.warning_default(),
+                            "Automatically updated to fresh commit {}",
```
nit: It's odd that warning is printed only for (less scary) `RecoveredSnapshot(_)` case, though I understand that's probably because we've printed lengthy status message for the `RecoveredMissingOp` case.

Maybe we can move all status/warning messages to `recover_stale_working_copy()` function instead of returning `StaleWorkingCopyStatus`? I thought `workspace_helper()` would call `recover_stale_working_copy()` unconditionally, but actually we check the snapshot error first. So I think it's safe to print `"Nothing to do .."` there.

yuja replied:
Looks better, thanks. Maybe we can insert a warning saying "recovering from stale working copy automatically..." or something, but it might be too verbose.

torquestomp replied:
I'll leave this as-is for now, we can iterate on it after there's some usage.
